src/redprl/refiner_types.fun

```diff
@@ -1502,6 +1508,175 @@ struct
+        val goalF = makeMem (H @> (z, AJ.TRUE (c, l, kApex))) (VarKit.rename (z, x) fx, (a, L.top, K.top))
+        val goalG = makeMem (H @> (z, AJ.TRUE (c, l, kApex))) (VarKit.rename (z, y) gy, (b, L.top, K.top))
+
+        val goalCohF = makeEq H ((substVar (m0, x) fx, fm0), (a, L.top, K.top))
+        val goalCohG = makeEq H ((substVar (m0, y) gy, gm0), (b, L.top, K.top))
+      in
+        |>: goalC >: goalA >: goalB >: goalCohF >: goalCohG >: goalF >: goalG #> (H, trivial)
+      end
+
+    fun EqFCom alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Pushout.EqFCom"
+        val H >> AJ.EQ ((tm0, tm1), (ty, l, k)) = jdg
+        val Syn.PUSHOUT (a, b, c, (x, fx), (y, gy)) = Syn.out ty
+        val Syn.FCOM args0 = Syn.out tm0
+        val Syn.FCOM args1 = Syn.out tm1
+
+        val (kEnd, kApex) = kindConstraintOnEndsAndApex k
+        val goalA = makeType H (a, l, kEnd)
+        val goalB = makeType H (b, l, kEnd)
+        val goalC = makeType H (c, l, kApex)
+
+        val z = alpha 0
+        val goalF = makeMem (H @> (z, AJ.TRUE (c, l, kApex))) (VarKit.rename (z, x) fx, (a, L.top, K.top))
+        val goalG = makeMem (H @> (z, AJ.TRUE (c, l, kApex))) (VarKit.rename (z, y) gy, (b, L.top, K.top))
+      in
+        |>: goalF >: goalG >: goalA >: goalB >: goalC
+         >:+ ComKit.genEqFComGoals H z (args0, args1) (ty, L.top, K.top)
+        #> (H, trivial)
+      end
+
+    fun Elim z alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Pushout.Elim"
+        val H >> AJ.TRUE (dz, l, k) = jdg
+        (* pushout-rec(FCOM) steps to COM *)
+        val k = K.meet (k, K.COM)
+        (* for now we ignore the kind and the level in the context *)
+        val AJ.TRUE (ty, _, _) = Hyps.lookup H z
+        val Syn.PUSHOUT (a, b, c, (x, fx), (y, gy)) = Syn.out ty
+
+        (* We need to kind-check cz because of FCOM
+         * This goal is made (explicitly) unconditional to simplify tactic writing
+         *)
+        val goalKind = makeType H (dz, l, k)
+
+        val m = alpha 0
+        val mtm = VarKit.toExp m
+
+        (* left branch *)
+        fun dleft tm = substVar (Syn.into (Syn.LEFT tm), z) dz
+        val (goalL, holeL) = makeTrue (H @> (m, AJ.TRUE (a, L.top, K.top))) (dleft mtm, L.top, K.top)
+
+        (* right branch *)
+        fun dright tm = substVar (Syn.into (Syn.RIGHT tm), z) dz
+        val (goalR, holeR) = makeTrue (H @> (m, AJ.TRUE (b, L.top, K.top))) (dright mtm, L.top, K.top)
+
+        (* glue branch *)
+        val v = alpha 1
+        val vtm = VarKit.toDim v
+        val fm = substVar (mtm, x) fx
+        val gm = substVar (mtm, y) gy
+        val glue = Syn.into @@ Syn.GLUE (vtm, mtm, fm, gm)
+        val dglue = substVar (glue, z) dz
+        val Hglue = H @> (v, AJ.TERM O.DIM) @> (m, AJ.TRUE (c, L.top, K.top))
+        val (goalG, holeG) = makeTrue Hglue (dglue, L.top, K.top)
+
+        (* coherence *)
+        val g0 = substVar (Syn.intoDim 0, v) holeG
+        val lf = substVar (fm, m) holeL
```

Comment:
@jonsterling Is it okay to do this substitution without worrying about non-terminating recursion?

Reply:
@favonia I think that it is fine, because substitution doesn't make any attempt to repeat itself until quiescence; it just replaces each occurrence of the original term with what you specify.